### PR TITLE
Implement SFQ(D) for query

### PR DESCRIPTION
**Pull request by kvinwang**

### CODE REVIEW

File: crates/phala-fair-queue/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+    }
+
+    pub fn purge_inactive_flows(&self, duration: Duration) {
+        self.inner.lock().unwrap().purge_inactive_flows(duration);
+    }
+}
+
+struct Flow {
+    previous_finish_tag: VirtualTime,
+    average_cost: VirtualTime,
```
[the remark this answers is not in the document]
Comment (h4x3rotab) [reply crates/phala-fair-queue/src/lib.rs:80]: Isn't it the EMA duration of the execution time? In my understanding only the time divided by the weight is virtual time.

[the remark this answers is not in the document]
Comment (kvinwang) [reply crates/phala-fair-queue/src/lib.rs:80]: That depends on how you define the VirtualTime. As it's name, it indicate the execution cost (duration of the execution in current implementation). In the future, we would normalize other resources (such as IO bytes) into the cost which could not be expressed by Duration.

**Comment by h4x3rotab** [comment]
SideVM queries don't block the full CPU. Should we account its execution time differently?

**Comment by kvinwang** [comment]
> SideVM queries don't block the full CPU. Should we account its execution time differently?

I found it is not easy to do it in this PR. Let me leave an [issue](https://github.com/Phala-Network/phala-blockchain/issues/885) there and resolve it later.

**Comment by kvinwang** [comment]
> SideVM queries don't block the full CPU. Should we account its execution time differently?

Now, the sidevm query bypass the fair queue totally. Let's account it in the future in Sidevm's shechuler.

**Review by h4x3rotab — APPROVED** [review]
LGTM

**Pull request merged**